src/librustc_codegen_llvm/common.rs

```diff
@@ -260,6 +261,7 @@ impl ConstMethods<'tcx> for CodegenCx<'ll, 'tcx> {
                     Some(GlobalAlloc::Function(fn_instance)) => self.get_fn_addr(fn_instance),
                     Some(GlobalAlloc::Static(def_id)) => {
                         assert!(self.tcx.is_static(def_id));
+                        assert!(!self.tcx.has_attr(def_id, sym::thread_local));
```

Comment:
You added this to a lot of places that *fetch* from the `alloc_map`; any reason you didn't add it at the place where things get inserted?

Reply:
the insertion functions themselves don't have access to the `tcx`. I could do it at the call sites, but since the only interesting call site is the one in librustc_mir_build, and that one already checks for thread locals and explicitly does *not* insert, an additional assertion would make no sense.

Reply:
I could use thread locals in `create_static_alloc`, but that seems odd and overkill.

Alternatively I can make the `AllocMap` private to `TyCtxt` and add functions to `TyCtxt` that forward to the `AllocMap`. That will cause a lot of churn, so I'd rather not do it in this PR.

Reply:
A comment in `create_static_alloc` sounds enough.

Reply:
I did the full thing in https://github.com/rust-lang/rust/pull/71508

Reply:
You mean, adding a comment? I cannot see it there, am I looking in the wrong place?

Reply:
No, I mean the forwarding of the functions. Once the other PR is merged, I can rebase this PR and add the assertion in `create_static_alloc`

Reply:
Oh I see.
Presumably, `alloc_map.create_static_alloc` also still exists though, and its doc comment should ideally mention this condition as well.

Reply:
It does not. There's only the `TyCtxt` method, which forwards to `reserve_and_set_dedup`, which is the place that will actually be doing the checking.